client/mm/mm.go

```diff
@@ -979,47 +1003,155 @@ func (m *MarketMaker) handleNotification(n core.Notification) {
 	}
 }
 
+// loadAndConnectCEX initializes the centralizedExchange if required, and
+// connects if not already connected.
+func (m *MarketMaker) loadAndConnectCEX(cfg *CEXConfig) (*centralizedExchange, *dex.ConnectionMaster, error) {
+	c, err := m.loadCEX(m.ctx, cfg)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error loading CEX: %w", err)
+	}
+	var cm *dex.ConnectionMaster
+	c.mtx.Lock()
+	if c.cm == nil || !c.cm.On() {
+		cm = dex.NewConnectionMaster(c)
+		c.cm = cm
+	} else {
+		cm = c.cm
+	}
+	c.mtx.Unlock()
+	if !cm.On() {
+		if err = cm.Connect(m.ctx); err != nil {
+			return nil, nil, fmt.Errorf("failed to connect to CEX: %v", err)
+		}
+	}
+	return c, cm, nil
+}
+
+// connectedCEX returns the connected centralizedExchange, initializing and
+// connecting if required.
+func (m *MarketMaker) connectedCEX(cexName string) (*centralizedExchange, error) {
+	cfg := m.cexConfig(cexName)
+	if cfg == nil {
+		return nil, fmt.Errorf("CEX %q not known", cexName)
+	}
+	cex, _, err := m.loadAndConnectCEX(cfg)
+	return cex, err
+}
+
+// initCEXConnections initializes and connects to the specified cexes.
 func (m *MarketMaker) initCEXConnections(cfgs []*CEXConfig) (map[string]libxc.CEX, map[string]*dex.ConnectionMaster) {
 	cexes := make(map[string]libxc.CEX)
 	cexCMs := make(map[string]*dex.ConnectionMaster)
 
 	for _, cfg := range cfgs {
-		if _, found := cexes[cfg.Name]; !found {
-			logger := m.log.SubLogger(fmt.Sprintf("CEX-%s", cfg.Name))
-			cex, err := libxc.NewCEX(cfg.Name, cfg.APIKey, cfg.APISecret, logger, dex.Simnet)
-			if err != nil {
-				m.log.Errorf("Failed to create %s: %v", cfg.Name, err)
-				continue
-			}
+		c, cm, err := m.loadAndConnectCEX(cfg)
+		if err != nil {
+			m.log.Errorf("Failed to create %s: %v", cfg.Name, err)
+			continue
+		}
+		cexes[c.Name] = c.CEX
+		cexCMs[c.Name] = cm
+	}
 
-			cm := dex.NewConnectionMaster(cex)
-			err = cm.Connect(m.ctx)
-			if err != nil {
-				m.log.Errorf("Failed to connect to %s: %v", cfg.Name, err)
-				continue
+	return cexes, cexCMs
+}
+
+// loadCEX initializes the cex if required and returns the centralizedExchange.“
+func (m *MarketMaker) loadCEX(ctx context.Context, cfg *CEXConfig) (*centralizedExchange, error) {
+	m.cexMtx.Lock()
+	defer m.cexMtx.Unlock()
+	var success bool
+	if cex := m.cexes[cfg.Name]; cex != nil {
+		if cex.Name == cfg.Name && cex.APISecret == cfg.APISecret {
+			return cex, nil
+		}
+		// New credentials. Delete the old cex.
+		defer func() {
+			if success {
+				cex.mtx.Lock()
+				cex.cm.Disconnect()
+				cex.cm = nil
+				cex.mtx.Unlock()
 			}
+		}()
+	}
+	logger := m.log.SubLogger(fmt.Sprintf("CEX-%s", cfg.Name))
+	cex, err := libxc.NewCEX(cfg.Name, cfg.APIKey, cfg.APISecret, logger, dex.Simnet)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create CEX: %v", err)
+	}
+	mkts, err := cex.Markets(ctx)
```

Comment:
When `Markets` is called here, the CEX is not yet connected. In the Binance library, the markets are loaded only when `Connect` is called, so nothing is returned here.